### PR TITLE
[Python] remove singleton

### PR DESCRIPTION
**Pull request by baartosz**
### PR checklist

- [X] Read the [contribution guildelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [X] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [X] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

Resolves https://github.com/swagger-api/swagger-codegen/issues/4986. Configuration class is not singleton anymore, similarly to PHP version. PR is to 2.3 branch as it changes a bit how client object is instantiated:
```
        config = petstore_api.Configuration()
        config.host = 'http://localhost/'
        api = PetApi(petstore_api.ApiClient(config))
        pet = api.get_pet_by_id(10)
```

### CODE REVIEW

**Comment by wing328** [comment]
@baartosz thanks for the PR.

cc @scottrw93 @frol @taxpon @scop, please review when you've time.

**Comment by frol** [comment]
This sounds like a great change! I am reviewing the implementation now.

@wing328 Thanks for pinging me.

File: samples/client/petstore/python/petstore_api/api_client.py [hunk 1]
```diff
@@ -116,7 +113,7 @@ def __call_api(self, resource_path, method,
                                                     collection_formats)
             for k, v in path_params:
                 resource_path = resource_path.replace(
-                    '{%s}' % k, quote(str(v), safe=''))  # no safe chars, encode everything
+                    '{%s}' % k, quote(str(v)))
```
Comment (frol) [inline samples/client/petstore/python/petstore_api/api_client.py]: @wing328 Why aren't these PRs merged into 2.3.0?

* https://github.com/swagger-api/swagger-codegen/commit/6a080d3a0c26015262120cdd9e7f4f2feeb59b23
* https://github.com/swagger-api/swagger-codegen/commit/c366f139e28b4b67f11a51807580f897afccc305

Comment (frol) [inline samples/client/petstore/python/petstore_api/api_client.py]: This change should be reverted to the upstream version in any case since it is irrelevant to this PR (and the former variant with `safe=''` is the correct one). Hopefully, the mentioned commits will land into 2.3.0 branch soon.

**Comment by frol** [comment]
@baartosz Great enhancement to the Python client! Thank you.

May I ask you to also update the README template to include the proper example of configuration?

File: modules/swagger-codegen/src/main/resources/python/api.mustache [hunk 2]
```diff
@@ -24,13 +24,9 @@ class {{classname}}(object):
     """
 
     def __init__(self, api_client=None):
-        config = Configuration()
```
Comment (frol) [inline modules/swagger-codegen/src/main/resources/python/api.mustache]:
```python
from ..configuration import Configuration
```

This import is unnecessary now.

File: samples/client/petstore/python/petstore_api/api_client.py (hunk 1 above)
Comment (baartosz) [inline samples/client/petstore/python/petstore_api/api_client.py]: should I also revert other unrelated changes in regenerated samples? e.g. `fake_classname_tags123_api.py` renamed to `fake_classname_tags_123_api`

**Comment by baartosz** [comment]
@frol Thanks for reviewing this PR. I added requested changes.

**Comment by frol** [comment]
@wing328 This change looks great to me now.

@wing328 Please, pick the 6a080d3 and c366f13 commits into 2.3.0 line from master.

@baartosz Thank you for your work!

**Comment by wing328** [comment]
> @wing328 Please, pick the 6a080d3 and c366f13 commits into 2.3.0 line from master.

I'm in the process of syncing enhancements from master to 2.3.0. Will let you guys know if I need any help. Thanks for listing out commits (enhancements, bug fixes) that need to get into 2.3.0.

**Comment by baartosz** [comment]
Rebased with 2.3.0 to solve merge conflicts

**Comment by wing328** [comment]
@baartosz PR merged into master. Thanks for the huge enhancement.

**Comment by wing328** [comment]
## Upgrade Note ##

`Configuration` is now per `ApiClient` instead of singleton.

**Comment by jdevera** [comment]
I found this when I was going to report a problem. I don't know if you have seen this, but my generated Api classes all call the ApiClient constructor without parameters, which always forces the creation of a a new configuration object per Api and leaves out the positibility of passing down any configuration from the client code.

**Comment by jdevera** [comment]
Ok, just realised that what I have to do is create an ApiClient with my custom configuration and pass it to the different Api objects.

**Comment by kmxillo** [comment]
When I create in the same Class the different API Client Configurations It overrides the configuration even though being called from different APIs. 

```python 
config1 = petstore_api_1.Configuration()
config2 = petstore_api_2.Configuration()
``` 
After than I do:
```python
api1 = PetApi(petstore_api_1.ApiClient(config1))
api2= PetApi(petstore_api_2.ApiClient(config2))

pet = api1.get_pet_by_id(10)
pet = api2.get_pet_by_id(10)
```

When I comment out one of the configurations it works as expected. 

Why cant I have several configurations when Configuration Class is not anymore Singleton?

**Comment by wing328** [comment]
cc @taxpon @frol @mbohlool @cbornet

**Comment by frol** [comment]
@kmxillo It is not clear what gets wrong on your side. There should be no global configuration anymore with Codegen 2.3.0+ (I don't believe 2.2.x include the changes). I also suggest you try the master and 3.0.0 branches just for the sake of testing, they are pretty stable.

**Comment by kmxillo** [comment]
@frol  I can see that the 2.3.0+ or 3.0.0 behavior is different as 2.2.* Thanks for that. 

But as I described above, for instance, some configurations settings are overwritten.  I would really like to know why this occurs. Since I see this unexpected behavior I cannot accept that  the implementation was successfully done.

Unexpected Behaviour:
- I dont see urllib3.connectionpool - DEBUG: Starting new ... for INT when I switched debug. 

  => I have seen that petstore_api_2.Configuration() overrides config_api1.logger['urllib3_logger'].level WHY?

I cant state that calls are executed correctly if I see such wrong behaviour.

As you see in my output. I call two environments {INT,PREPROD}.  

**OUTPUT:** 
```
pythonapp - INFO: Executing Activation Services at: INT
pythonapp - INFO: {'basicAuth': {'type': 'basic'}}
DEBUG response body: Code 200
petApi.rest - DEBUG: response body: Successfully added.
pythonapp - INFO: Executing Activation Services at: PREPROD
pythonapp - INFO: {'basicAuth': {'type': 'basic'}}
DEBUG Starting new HTTPS connection (1): PREPROD.host
DEBUG Starting new HTTPS connection (1): PREPROD.host
urllib3.connectionpool - DEBUG: Starting new HTTPS connection (1): PREPROD.host
```
Here is how I initialized the **configurations**:

```
class ClientSettings(object):

    def __new__(self,system):

        #  Config for petstore_api_1
        # ########################
        config_api1 = petstore_api_1.Configuration()
        config_api1 .debug = True
        config_api1.host = 'INT.host'
        config_api1.username = username
        config_api1.password = password

        # Config for petstore_api_2
        # #######################
        config_api2 = petstore_api_2.Configuration()
        config_api2.host = 'PREPROD.host'
        config_api2.username = username
        config_api2.password = password

        configurations = {}
        configurations['api1'] = config_api1
        configurations['api2'] = config_api2

        return configuration
```

**Comment by frol** [comment]
@kmxillo Well, there is a definitely weird implementation around logging:

1. Configuration.__init__ sets `self.debug = False`: https://github.com/swagger-api/swagger-codegen/blob/master/samples/client/petstore/python/petstore_api/configuration.py#L83
2. `self.debug`, being a property, updates the level of the loggers: https://github.com/swagger-api/swagger-codegen/blob/master/samples/client/petstore/python/petstore_api/configuration.py#L164-L176

This doesn't make Configuration to be in a shared state, but it, indeed, does unexpected things with the logging configuration. I don't know how to fix that, but remove it.

**Comment by e-desouza** [comment]
I can confirm (in 2023!) that this does indeed work as expected (no singleton Configuration).
